File: leetcode/graph/505_The_Maze_II/codes/dijkstra.py
import heapq
from typing import List

class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        dist = [[float("inf") for _ in range(n)] for _ in range(m)]
        dist[start[0]][start[1]] = 0
        heap = [(0, start[0], start[1])]

        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        while heap:
            d, x, y = heapq.heappop(heap)
            if [x, y] == destination:
                return d
            for dx, dy in directions:
                nx, ny, steps = x, y, 0
                while 0 <= nx+dx < m and 0 <= ny+dy < n and maze[nx+dx][ny+dy] == 0:
                    nx += dx
                    ny += dy
                    steps += 1
                if d + steps < dist[nx][ny]:
                    dist[nx][ny] = d + steps
                    heapq.heappush(heap, (d + steps, nx, ny))
                    
        return -1

File: leetcode/graph/505_The_Maze_II/codes/dijkstra.cpp
#include <vector>
#include <queue>
#include <climits>
using namespace std;

class Solution {
public:
    int shortestDistance(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {
        int m = maze.size(), n = maze[0].size();
        vector<vector<int>> dist(m, vector<int>(n, INT_MAX));
        dist[start[0]][start[1]] = 0;
        priority_queue<vector<int>, vector<vector<int>>, greater<>> pq;
        pq.push({0, start[0], start[1]});
        
        vector<vector<int>> dirs = {{-1, 0},{1, 0},{0, -1},{0, 1}};
        
        while (!pq.empty()) {
            vector<int> cur = pq.top(); 
            int d = cur[0], x = cur[1], y = cur[2];
            pq.pop();
            if (x == destination[0] && y == destination[1]) return d;
            for (auto& dir : dirs) {
                int nx = x, ny = y, steps = 0;
                while (nx + dir[0] >= 0 && nx + dir[0] < m &&
                       ny + dir[1] >= 0 && ny + dir[1] < n &&
                       maze[nx + dir[0]][ny + dir[1]] == 0) {
                    nx += dir[0];
                    ny += dir[1];
                    steps++;
                }
                if (d + steps < dist[nx][ny]) {
                    dist[nx][ny] = d + steps;
                    pq.push({d + steps, nx, ny});
                }
            }
        }
        return -1;
    }
};
